### PR TITLE
Browser Integration code cleanup

### PR DESCRIPTION
**Pull request by varjolintu**
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Explain large or complex code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" )
Cleanups the browser integration code to use structs whenever necessary. Decreases the amount of function parameters and adds some frequently used functions to the structs itself.

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Manually.

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Refactor (significant modification to existing code)

### CODE REVIEW

File: src/browser/BrowserAction.h [hunk 1]
```diff
@@ -52,6 +76,7 @@ class BrowserAction
     QJsonObject buildResponse(const QString& action, const QJsonObject& message, const QString& nonce);
     QJsonObject getErrorReply(const QString& action, const int errorCode) const;
     QJsonObject decryptMessage(const QString& message, const QString& nonce);
+    BrowserRequest getBrowserRequest(const QJsonObject& json);
```
Comment (droidmonkey) [inline src/browser/BrowserAction.h]: `toBrowserRequest` or maybe just `decodeRequest`. Parse and Unpack are possible verbs as well. 

Another take on this is that when a request is received, the internals of BrowserService just immediately unpack it into the struct and pass the struct around. Then the only function that takes json is the read function. 

You'd likewise have a response struct as well.

Comment (varjolintu) [reply src/browser/BrowserAction.h]: Sure. I can change the name and look the other suggestions too. I had an idea that this PR still needs some work.

Comment (varjolintu) [reply src/browser/BrowserAction.h]: The only way to use the struct directly would be to divide the requests in two groups: those which need the decrypted content and association, and those which do not. The first group would handle the struct.

**Comment by varjolintu** [comment]
Changes done that revert the asynchronous stuff.

**Comment by varjolintu** [comment]
Maybe this could be merged after 2.7.5? I have already another PR coming that uses this branch as a base, fixing a few annoying bugs with the Access Confirm Dialog.

**Comment by droidmonkey** [comment]
This'll stay in develop not the 2.7.x line

File: src/browser/BrowserAction.cpp [hunk 2]
```diff
@@ -127,171 +136,150 @@ QJsonObject BrowserAction::handleChangePublicKeys(const QJsonObject& json, const
...
-    const QString auth = decrypted.value("httpAuth").toString();
+    const auto id = browserRequest.getString("id");
+    const auto formUrl = browserRequest.getString("submitUrl");
+    const auto auth = browserRequest.getString("httpAuth");
     const bool httpAuth = auth.compare(TRUE_STR) == 0;
 
-    const QJsonArray users = browserService()->findMatchingEntries(id, siteUrl, formUrl, "", keyList, httpAuth);
+    EntryParameters entryParameters;
+    entryParameters.dbid = id;
+    entryParameters.hash = browserRequest.hash, entryParameters.siteUrl = siteUrl;
```
Comment (droidmonkey) [inline src/browser/BrowserAction.cpp]: Is this comma intentional?

File: src/browser/BrowserAction.cpp [hunk 3]
```diff
@@ -127,171 +136,150 @@ QJsonObject BrowserAction::handleChangePublicKeys(const QJsonObject& json, const
...
+    entryParameters.formUrl = formUrl;
+
+    const QJsonArray users = browserService()->findEntries(entryParameters, keyList, httpAuth);
```
Comment (droidmonkey) [inline src/browser/BrowserAction.cpp]: recommend `const auto` here

File: src/browser/BrowserAction.cpp [hunk 4]
```diff
@@ -305,108 +293,102 @@ QJsonObject BrowserAction::handleGeneratePassword(QLocalSocket* socket, const QJ
         return errorReply;
     }
 
-    browserService()->showPasswordGenerator(socket, incrementedNonce, m_clientPublicKey, m_secretKey);
-    return QJsonObject();
+    KeyPairMessage keyPairMessage = {socket, browserRequest.incrementedNonce, m_clientPublicKey, m_secretKey};
```
Comment (droidmonkey) [inline src/browser/BrowserAction.cpp]: I believe you can drop the equals sign to directly init. Doing it this way creates a temporary KeyPairMessage then copies it to the variable (I think).

File: src/browser/BrowserAction.cpp [hunk 5]
```diff
@@ -305,108 +293,102 @@ QJsonObject BrowserAction::handleGeneratePassword(QLocalSocket* socket, const QJ
...
+    const auto command = browserRequest.getString("action");
+    if (!command.isEmpty() && command.compare(BROWSER_REQUEST_LOCK_DATABASE) == 0) {
         browserService()->lockDatabase();
 
-        const QString newNonce = browserMessageBuilder()->incrementNonce(nonce);
-        QJsonObject message = browserMessageBuilder()->buildMessage(newNonce);
-
-        return buildResponse(action, message, newNonce);
+        auto message = browserMessageBuilder()->buildMessage(browserRequest.incrementedNonce);
+        return buildResponse(action, message, browserRequest.incrementedNonce);
```
Comment (droidmonkey) [inline src/browser/BrowserAction.cpp]: Why do you pass the incrementedNonce into buildResponse when it is already a part of `message`?

File: src/browser/BrowserAction.cpp [hunk 6]
```diff
@@ -127,171 +136,150 @@ QJsonObject BrowserAction::handleChangePublicKeys(const QJsonObject& json, const
...
-    QJsonObject message = browserMessageBuilder()->buildMessage(incrementedNonce);
+    auto message = browserMessageBuilder()->buildMessage(browserRequest.incrementedNonce);
     message["count"] = users.count();
     message["entries"] = users;
-    message["hash"] = hash;
+    message["hash"] = browserRequest.hash;
     message["id"] = id;
 
-    return buildResponse(action, message, incrementedNonce);
+    return buildResponse(action, message, browserRequest.incrementedNonce);
```
Comment (droidmonkey) [inline src/browser/BrowserAction.cpp]: This pattern is all over the new code base: buildMessage, maybe add message params, build response passing in message. I feel like we can simplify this greatly:

buildResponse(action, nonce, QMap<QString, QVariant> params);

then you call it like:

buildResponse(action, nonce, parammap);

use `message[parammap.key] = QJsonValue::fromVariant(parammap.value);`

**Review by droidmonkey — CHANGES_REQUESTED** [review]

File: src/browser/BrowserAction.cpp (hunk 5 above)
Comment (varjolintu) [reply src/browser/BrowserAction.cpp]: Currently it's needed for both (message verification on the extension side), but I'm about to drop it in the future. Not just yet. I could use `message["nonce"]` instead.

File: src/browser/BrowserAction.cpp (hunk 6 above)
Comment (varjolintu) [reply src/browser/BrowserAction.cpp]: This should work fine:
```
typedef QMap<QString, QVariant> Parameters;

...

const Parameters params {
    { "name", newGroup["name"] },
    { "uuid", newGroup["uuid"] }
};
return buildResponse(action, browserRequest.incrementedNonce, params);

... inside buildResponse():

Parameters::const_iterator i;
for (i = params.constBegin(); i != params.constEnd(); ++i) {
    message[i.key()] = QJsonValue::fromVariant(i.value());
}
```

Comment (droidmonkey) [reply src/browser/BrowserAction.cpp]: Yah that is great! Chops down some boiler plate code

**Review by droidmonkey — APPROVED** [review]

**Comment by droidmonkey** [comment]
Man that looks great now!

**Comment by varjolintu** [comment]
I could add a couple tests for the message creation, just in case.

**Comment by droidmonkey** [comment]
Yah that's a good idea, really to test expected parameters and the nonce matches would be fine.

**Comment by varjolintu** [comment]
Yeah. It's actually quite easy to make a const `QJsonObject` just by mistake and add parameters there, which are not there at runtime. And the compiler doesn't complain about such errors.

EDIT: Made sure that I'm not doing those kind of errors..

**Pull request merged**